lib/Backend/Vault/Kv/Internal.hs

```diff
@@ -0,0 +1,342 @@
+--     "custom_metadata":
+--     { "owner": "jdoe",
+--       "mission_critical": "false"
+--     },
+--     "deletion_time": "",
+--     "destroyed": false,
+--     "version": 2
+--   }
+-- }
+data ReadSecret =
+  ReadSecret
+  { _secret :: HS.HashMap T.Text T.Text
+  , _rCustomMetadata :: HS.HashMap T.Text T.Text
+  , _createdTime :: T.Text
+  , _deletionTime :: T.Text
+  , _destroyed :: Bool
+  , _version :: Int
+  }
+  deriving (Show)
+makeLenses ''ReadSecret
+
+
+-- |
+-- A type defining the request to the Vault servet at endpoint
+-- '/v1/<MOUNT>/data/<PATH>'.
+-- For example:
+--
+-- { "options":
+-- , { "cas": 0
+--   }
+-- , "data":
+--   { "foo": "bar"
+--   , "zip": "zap"
+--   }
+-- }
+data PostSecret =
+  PostSecret
+  { _cas :: Maybe Int
+  , _pdata :: HS.HashMap T.Text T.Text
+  }
+  deriving (Show)
+type PatchSecret = PostSecret
+makeLenses ''PostSecret
+
+
+-- |
+-- A type defining the request to the Vault servet at endpoint
+-- '/v1/<MOUNT>/metadata/<PATH>'.
+-- For example:
+--
+-- { "max_versions": 5
+-- , "cas_required": false
+-- , "delete_version_after": "40m"
+-- , "custom_metadata":
+--   { "example": ":)"
+--   }
+-- }
+data UpdateMetadata =
+  UpdateMetadata
+  { _maxVersions :: Maybe Int
+  , _casRequired :: Maybe Bool
+  , _deleteVersionAfter :: Maybe T.Text
+  , _uCustomMetadata :: HS.HashMap T.Text T.Text
+  }
+  deriving (Show)
+makeLenses ''UpdateMetadata
+
+-- Overloaded Lens accessors
+
+class DataLens a b where
+  ddata :: Lens' a b
+
+instance a ~ b => DataLens (KvResponse a) b where
+  ddata = kdata
+
+instance DataLens PostSecret (HS.HashMap T.Text T.Text) where
+  ddata = pdata
+
+class CustomMetadataLens a where
+  customMetadata :: Lens' a (HS.HashMap T.Text T.Text)
+
+instance CustomMetadataLens ReadSecret where
+  customMetadata = rCustomMetadata
+
+instance CustomMetadataLens UpdateMetadata where
+  customMetadata = uCustomMetadata
+
+-- JSON serialization/deserialization, logically some ADTs need to only be deserialized,
+-- others serialized, but never both.
+
+instance A.FromJSON ListSecrets where
+  parseJSON = A.withObject "ListSecrets" $ \o ->
+    ListSecrets <$> o .: "keys"
+
+instance A.FromJSON ReadSecret where
+  parseJSON = A.withObject "ReadSecret" $ \o -> do
+    metadata <- o .: "metadata"
+    ReadSecret
+      <$> o .: "data"
+      <*> (metadata .: "custom_metadata" <|> pure (HS.fromList []))
+      <*> metadata .: "created_time"
+      <*> metadata .: "deletion_time"
+      <*> metadata .: "destroyed"
+      <*> metadata .: "version"
+
+instance A.FromJSON a => A.FromJSON (KvResponse a) where
+  parseJSON = A.withObject "KvResponse" $ \o ->
+    KvResponse
+    <$> o .: "request_id"
+    <*> o .: "lease_id"
+    <*> o .: "renewable"
+    <*> o .: "lease_duration"
+    <*> (o .: "data" >>= A.parseJSON)
+    <*> pure Nothing
+    <*> pure Nothing
+    <*> pure Nothing
+
+removeNull :: [AT.Pair] -> [AT.Pair]
+removeNull = filter (\(_, v) -> case v of A.Null -> False ; _ -> True)
```

Comment:
```suggestion
removeNull = filter \case
  (_, A.Null) -> False
  _ -> True
```

Reply:
`BlockArgument` is nice

Reply:
You mean `LambdaCase` :P

Reply:
_And_ `BlockArguments`, because otherwise it will be `filter $ \case` :D

Reply:
Done

Reply:
> And BlockArguments, because otherwise it will be filter $ \case :D

Oh right, I'm used to associating `BlockArguments` with `do`, but you're absolutely right :D